### DISCUSSION TITLE
How to generate a MicroPython QSTR from a C macro?

### DISCUSSION THREAD

**Discussion by pablogventura** [micropython/micropython#10966] [Using MicroPython]
I'm using MicroPython and I would like to generate a QSTR that contains the contents of a macro. For example, I have the following macro:

`#define MY_MACRO "hello"`

How can I generate a QSTR that has the value "hello" and is called MP_QSTR_MY_MACRO?

When I use MP_QSTR_MY_MACRO, that QSTR contains "MY_MACRO" instead of "hello".

**Comment by dlech** [comment]
Just use `MP_QSTR_hello` without `#define`ing it.

There is a tool that automatically creates this macro based on the pattern `MP_QSTR_(\w+)`.

**Reply by pablogventura** [reply]
Thanks for your answer! Unfortunately my problem is that the macro I need to use to generate the QSTR is generated at compile time, in the manner of `__USER__`.

**Reply by dlech** [reply]
Then you should be able to create a generated file with lines like:

```
Q(hello)
```

and add that file to `QSTR_DEFS` in your specific Makefile. 

The STM32 port has something like this already, e.g. for generating pin names from a config file (i.e. search for `QSTR_DEFS += qstrdefsport.h $(QSTR_GENERATED_HEADERS)`).

**Reply by pablogventura** [reply]
Thank you for your response! Unfortunately the problem continues. In particular I have a macro that returns the git branch (GIT_BRANCH). If I generate the QSTR with `Q(GIT_BRANCH)` it generates `QDEF(MP_QSTR_git_branch, 59348, 10, "git_branch")`. So it does not work for me.

I would need to be able to declare a QSTR whose name is static, but its value comes from a macro at compile time.

**Reply by dlech** [reply]
Does it actually have to be a QSTR? MicroPython uses `MP_DEFINE_STR_OBJ` instead for things like the git tag.
